### PR TITLE
Introduce `ckpt_path="hpc"` keyword for checkpoint loading

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

Implements first two points from https://github.com/Lightning-AI/lightning/issues/13773#issuecomment-1227409044

### Does your PR introduce any breaking changes? If yes, please list them.

HPC checkpoints are now *NOT* being loaded first as a default. Instead, they are specifically loaded if a user passes keyword `"hpc"`. However, if a user is running with `SLURMEnvironment` and passes `ckpt_path=None` (which is a default value) and an HPC checkpoint is found, then it is automatically selected.

I'm not sure if we need more tests around this, if yes, I will implement them

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj @justusschock @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -75,8 +76,7 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None) -> None:
         3. from `checkpoint_path` file if provided
         4. don't restore
         """
-        self.resume_checkpoint_path = self._hpc_resume_path or checkpoint_path
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:78]: Okay, so this is the breaking change you mention in the changelog. No longer do we auto detect the file, and only look at if if ckpt_path="hpc". LGTM

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -105,6 +105,9 @@ def _set_ckpt_path(
         ft_checkpoints = [cb for cb in self.trainer.callbacks if isinstance(cb, _FaultToleranceCheckpoint)]
         fn = state_fn.value
 
+        if ckpt_path is None and SLURMEnvironment.detect() and self._hpc_resume_path is not None:
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:108]: Do you think it's needed to check for `SLURMEnvironment.detect()`?

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:78]: Would you mind adjusting the docstring of the first item above: "1. from HPC weights if found" -> "1. from HPC if checkpont_path has the special name 'hpc'."

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -105,6 +105,9 @@ def _set_ckpt_path(
         ft_checkpoints = [cb for cb in self.trainer.callbacks if isinstance(cb, _FaultToleranceCheckpoint)]
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:105]: You split this up, I would put this together with the `if` that uses it

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 2 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:108]: I think it's safer in case somebody did a rsync of a SLURM filesystem, but this is an extreme edge case

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -175,6 +178,9 @@ def _set_ckpt_path(
                 return None
             ckpt_path = max(candidates_ts.keys(), key=partial(operator.getitem, candidates_ts))
 
+        if ckpt_path == "hpc":
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:181]: Should be an `elif`, no? Same for the checks above

File: tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py [hunk 5]
```diff
@@ -75,13 +76,6 @@ def test_hpc_restore_attempt(tmpdir):
         assert param.abs().sum() > 0
         torch.nn.init.constant_(param, 0)
 
-    # case 2: explicit resume path provided, restore hpc anyway
-    trainer = Trainer(default_root_dir=tmpdir, max_steps=3)
-    trainer.fit(model, ckpt_path="not existing")
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:80]: Before this got ignored, now, it would raise an error that "not existing" does not exist. Correct?

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md [hunk 6]
```diff
@@ -51,6 +51,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added a warning when the model passed to `LightningLite.setup()` does not have all parameters on the same device ([#14822](https://github.com/Lightning-AI/lightning/pull/14822))
 
 
+- Introduce `ckpt_path="hpc"` keyword for checkpoint loading ([#14911](https://github.,com/Lightning-AI/lightning/pull/14911))
```
Comment (carmocca) [inline src/pytorch_lightning/CHANGELOG.md:54]: Can you also update the docstrings for the `ckpt_path` params of `trainer.fit/validate/...`

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:181]: I don't think it really matters, since `ckpt_path` will be overwritten within those checks and I honestly find it a bit easier to follow. But, sure, let's make it a bunch of `elif`s, I don't necessarily hold strong opinion on that.

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 2 above)
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:108]: I agree with @carmocca, as far as I understand HPC checkpoints are useless outside of SLURM environment

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:105]: I'm not sure I follow? It's being used in multiple `if`s, first usage is on line 111

File: tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py (hunk 5 above)
Comment (otaj) [reply tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:80]: That is correct. For clarity, I restored this test case with expecting annnnnnn exception

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:107]: https://github.com/Lightning-AI/lightning/pull/14911/commits/89a1fc8a52876423b56970e87912a53d9b4a3ea5

**Comment by otaj** [comment]
Those pesky docstrings :joy:. I updated them, @carmocca, let me know if something more should be said there.

**Review by tchaton — APPROVED** [review]
LGTM !
